zero/src/prover.rs

```diff
@@ -175,32 +186,48 @@ impl BlockProverInput {
 
         let block_number = self.get_block_number();
         info!("Testing witness generation for block {block_number}.");
-
-        let block_generation_inputs = trace_decoder::entrypoint(
+        let (block_generation_inputs, opt_struct_logs) = trace_decoder::entrypoint(
             self.block_trace,
             self.other_data,
+            self.struct_logs,
             batch_size,
             &mut DummyObserver::new(),
         )?;
-
         let seg_ops = ops::SegmentProofTestOnly {
             save_inputs_on_error,
             save_tries_on_error,
         };
 
-        let simulation = Directive::map(
-            IndexedStream::from(
-                block_generation_inputs
-                    .into_iter()
-                    .enumerate()
-                    .zip(repeat(max_cpu_len_log))
-                    .map(|((batch_index, txn_batch), max_cpu_len_log)| {
-                        (txn_batch, max_cpu_len_log, batch_index)
-                    }),
-            ),
-            &seg_ops,
-        );
-
+        let simulation = if let Some(struct_logs) = opt_struct_logs {
+            Directive::map(
+                IndexedStream::from(
+                    block_generation_inputs
+                        .into_iter()
+                        .enumerate()
+                        .zip(repeat(max_cpu_len_log))
+                        .zip(struct_logs)
+                        .map(
+                            |(((batch_index, txn_batch), max_cpu_len_log), struct_log)| {
+                                (txn_batch, max_cpu_len_log, batch_index, Some(struct_log))
+                            },
+                        ),
+                ),
+                &seg_ops,
+            )
+        } else {
+            Directive::map(
+                IndexedStream::from(
+                    block_generation_inputs
+                        .into_iter()
+                        .enumerate()
+                        .zip(repeat(max_cpu_len_log))
+                        .map(|((batch_index, txn_batch), max_cpu_len_log)| {
+                            (txn_batch, max_cpu_len_log, batch_index, None)
+                        }),
+                ),
+                &seg_ops,
+            )
+        };
```

Comment:
On multiple places you are duplicating code by checking the `opt_struct_logs`, just pass it through to the function.

Reply:
But here, I needed to iterate on the internal struct_logs vec (if there is any)...

Reply:
You could do this:
```
        let simulation = Directive::map(
            IndexedStream::from(
                block_generation_inputs
                    .clone()
                    .into_iter()
                    .enumerate()
                    .zip(repeat(prover_config.max_cpu_len_log))
                    .zip(opt_struct_logs.map_or(vec![vec![]; block_generation_inputs.len()], |x| x))
                    .map(|(((batch_index, inputs), max_cpu_len_log), struct_log)| {
                        (inputs, max_cpu_len_log, batch_index, Some(struct_log))
                    }),
            ),
            &seg_ops,
        );
```